sw/device/tests/dif_gpio/dif_gpio_test.cc

```diff
@@ -0,0 +1,85 @@
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "mock_mmio.h"
+
+extern "C" {
+#include "gpio_regs.h"
+#include "sw/device/lib/dif/dif_gpio.h"
+}
+
+// This is a temporary measure to work around the dependency on
+// |GPIO0_BASE_ADDR| in |GPIO_*()| family of macros in gpio_regs.h.
+// See lowrisc/opentitan#1118.
+static const uintptr_t GPIO0_BASE_ADDR = 0x0U;
+#define DEFINE_GPIO_REG_CONSTANT(NAME) \
+  static const ptrdiff_t GPIO_##NAME##_REG_OFFSET = GPIO_##NAME(0);
+DEFINE_GPIO_REG_CONSTANT(INTR_STATE)
+DEFINE_GPIO_REG_CONSTANT(INTR_ENABLE)
+DEFINE_GPIO_REG_CONSTANT(INTR_TEST)
+DEFINE_GPIO_REG_CONSTANT(DATA_IN)
+DEFINE_GPIO_REG_CONSTANT(DIRECT_OUT)
+DEFINE_GPIO_REG_CONSTANT(MASKED_OUT_LOWER)
+DEFINE_GPIO_REG_CONSTANT(MASKED_OUT_UPPER)
+DEFINE_GPIO_REG_CONSTANT(DIRECT_OE)
+DEFINE_GPIO_REG_CONSTANT(MASKED_OE_LOWER)
+DEFINE_GPIO_REG_CONSTANT(MASKED_OE_UPPER)
+DEFINE_GPIO_REG_CONSTANT(INTR_CTRL_EN_RISING)
+DEFINE_GPIO_REG_CONSTANT(INTR_CTRL_EN_FALLING)
+DEFINE_GPIO_REG_CONSTANT(INTR_CTRL_EN_LVLHIGH)
+DEFINE_GPIO_REG_CONSTANT(INTR_CTRL_EN_LVLLOW)
+DEFINE_GPIO_REG_CONSTANT(CTRL_EN_INPUT_FILTER)
+#undef DEFINE_GPIO_REG_CONSTANT
+
+using ::testing::Return;
+
+// Equal to operator to be able to write EXPECT_EQ(expGpio, actGpio).
+bool operator==(const dif_gpio_t &lhs, const dif_gpio_t &rhs) {
+  return lhs.base_addr == rhs.base_addr;
+}
+
+// Test fixture that provides a |dif_gpio_t| instance and some convenience
+// methods to simplify tests.
+class DifGpioTest : public MmioMockTest {
+ protected:
+  DifGpioTest() : gpio_{reg32_from_addr(0xAABBCCDDU)} {}
+
+  // Helper methods to simplify tests. They return a reference to the mock
+  // so that expectations can be modified.
+  auto &EXPECT_WRITE(const uint32_t offset, const uint32_t val) {
+    return EXPECT_REG32_WRITE(gpio_.base_addr, offset, val);
+  }
+
+  auto &EXPECT_READ(const uint32_t offset) {
+    return EXPECT_REG32_READ(gpio_.base_addr, offset);
+  }
+
+  dif_gpio_t gpio_;
+};
+
+// Tests
+
+TEST_F(DifGpioTest, Init) {
+  EXPECT_WRITE(GPIO_INTR_ENABLE_REG_OFFSET, 0xF1F1F1F1U);
+  EXPECT_WRITE(GPIO_DIRECT_OE_REG_OFFSET, 0xE2E2E2E2U);
+  EXPECT_WRITE(GPIO_INTR_CTRL_EN_RISING_REG_OFFSET, 0xD3D3D3D3U);
+  EXPECT_WRITE(GPIO_INTR_CTRL_EN_FALLING_REG_OFFSET, 0xC4C4C4C4U);
+  EXPECT_WRITE(GPIO_INTR_CTRL_EN_LVLHIGH_REG_OFFSET, 0xB5B5B5B5U);
+  EXPECT_WRITE(GPIO_INTR_CTRL_EN_LVLLOW_REG_OFFSET, 0xA6A6A6A6U);
+  EXPECT_WRITE(GPIO_CTRL_EN_INPUT_FILTER_REG_OFFSET, 0x97979797U);
+
+  dif_gpio_config_t config{
+      /* base_addr = */ gpio_.base_addr,
+      /* interrupt_enable = */ 0xF1F1F1F1U,
+      /* direct_output_enable = */ 0xE2E2E2E2U,
+      /* interrupt_enable_rising_edge = */ 0xD3D3D3D3U,
+      /* interrupt_enable_falling_edge = */ 0xC4C4C4C4U,
+      /* interrupt_enable_level_high = */ 0xB5B5B5B5U,
+      /* interrupt_enable_level_low = */ 0xA6A6A6A6U,
+      /* input_filter_enable = */ 0x97979797U,
+  };
```

Comment:
Let's use designated initializers so we avoid this.  The style guide allows them for C++ even though they're an extension and not in the spec yet.

Reply:
Well, the style guide at https://google.github.io/styleguide/cppguide.html doesn't, though this may change in the future. I'm ok with adopting the GCC/Clang extension in our style guide, so long as we conform to the C++20 version (initializers must be in order, no array initializers).